akka-http-core/src/main/scala/akka/http/impl/engine/http2/hpack/HeaderCompression.scala

```diff
@@ -25,60 +25,59 @@ private[http2] object HeaderCompression extends GraphStage[FlowShape[FrameEvent,
-          os.reset()
-          if (result.size <= currentMaxFrameSize) push(eventsOut, HeadersFrame(streamId, endStream, endHeaders = true, result, prioInfo))
-          else {
-            val first = HeadersFrame(streamId, endStream, endHeaders = false, result.take(currentMaxFrameSize), prioInfo)
+    def onPull(): Unit = pull(eventsIn)
+    def onPush(): Unit = grab(eventsIn) match {
+      case ack @ SettingsAckFrame(s) =>
+        applySettings(s)
+        push(eventsOut, ack)
+      case ParsedHeadersFrame(streamId, endStream, kvs, prioInfo) =>
+        kvs.foreach {
+          case (key, value: String) =>
+            encoder.encodeHeader(os, key, value, false)
+          case (key, value) =>
+            throw new IllegalStateException(s"Didn't expect key-value-pair [$key] -> [$value](${value.getClass}) here.")
+        }
+        val result = ByteString.fromArrayUnsafe(os.toByteArray) // BAOS.toByteArray always creates a copy
+        os.reset()
+        if (result.size <= currentMaxFrameSize) push(eventsOut, HeadersFrame(streamId, endStream, endHeaders = true, result, prioInfo))
+        else {
+          val first = HeadersFrame(streamId, endStream, endHeaders = false, result.take(currentMaxFrameSize), prioInfo)
 
-            emit(eventsOut, first)
-            setHandler(eventsOut, new OutHandler {
-              var remainingData = result.drop(currentMaxFrameSize)
+          push(eventsOut, first)
+          setHandler(eventsOut, new OutHandler {
+            private var remainingData = result.drop(currentMaxFrameSize)
 
-              def onPull(): Unit = {
-                val thisFragment = remainingData.take(currentMaxFrameSize)
-                val rest = remainingData.drop(currentMaxFrameSize)
-                val last = rest.isEmpty
+            def onPull(): Unit = {
+              val thisFragment = remainingData.take(currentMaxFrameSize)
+              val rest = remainingData.drop(currentMaxFrameSize)
+              val last = rest.isEmpty
 
-                push(eventsOut, ContinuationFrame(streamId, endHeaders = last, thisFragment))
-                if (last) become(Idle)
-                else remainingData = rest
-              }
-            })
-          }
-      }
-
-      def applySettings(s: immutable.Seq[Setting]): Unit =
-        s foreach {
-          case Setting(SettingIdentifier.SETTINGS_HEADER_TABLE_SIZE, size) =>
-            log.debug("Applied SETTINGS_HEADER_TABLE_SIZE({}) in header compression", size)
-            // 'size' is strictly spoken unsigned, but the encoder is allowed to
-            // pick any size equal to or less than this value (6.5.2)
-            if (size >= 0) encoder.setMaxHeaderTableSize(os, size)
-            else encoder.setMaxHeaderTableSize(os, Int.MaxValue)
-          case _ => // ignore, not applicable to this stage
+              push(eventsOut, ContinuationFrame(streamId, endHeaders = last, thisFragment))
+              if (last) setHandler(eventsOut, logic)
```

Comment:
should we also `pull(eventsIn)` here?

Reply:
I think we wait again for the next `pull` which will then `pull(eventsIn)`.

Reply:
We sent something in the line above, so there should be pull.

Reply:
You're right, I'm not sure what I was thinking :)